### PR TITLE
config: async data access for local and remote data source

### PR DESCRIPTION
**Pull request by crazyxy**
Signed-off-by: crazyxy <yxyan@google.com>

Description: Add an async data source which supports fetching data from local and remote data source. The async data provider guarantees that data is access before `init manager` is ready.
Risk Level: Low
Testing: Unit test
Docs Changes: N/A
Release Notes: https://github.com/envoyproxy/envoy/compare/master...crazyxy:datasource?expand=1#diff-d8d3e33358b55e6c0466dd1bb5f40c79
Fixes #Issue: #7311

### CODE REVIEW

File: source/common/config/datasource.h [hunk 1]
```diff
@@ -25,6 +32,108 @@ std::string read(const envoy::api::v2::core::DataSource& source, bool allow_empt
...
+  LocalAsyncDataProvider(Init::Manager& manager, const envoy::api::v2::core::DataSource& source,
+                         bool allow_empty, Api::Api& api, AsyncDataSourceCb&& callback)
+      : init_target_("LocalAsyncDataProvider", [this, &source, allow_empty, &api, callback]() {
+          callback(DataSource::read(source, allow_empty, api));
+          init_target_.ready();
+        }) {
+    manager.add(init_target_);
+  }
+
+  ~LocalAsyncDataProvider() { init_target_.ready(); }
```
Comment (silentdai) [inline source/common/config/datasource.h]: Could you explain why destructor should have the side effect as target ready?

File: source/common/config/datasource.h [hunk 2]
```diff
@@ -25,6 +32,108 @@ std::string read(const envoy::api::v2::core::DataSource& source, bool allow_empt
...
+class RemoteAsyncDataProvider;
+using RemoteAsyncDataProviderPtr = std::unique_ptr<RemoteAsyncDataProvider>;
+
+class LocalAsyncDataProvider {
+public:
+  LocalAsyncDataProvider(Init::Manager& manager, const envoy::api::v2::core::DataSource& source,
+                         bool allow_empty, Api::Api& api, AsyncDataSourceCb&& callback)
+      : init_target_("LocalAsyncDataProvider", [this, &source, allow_empty, &api, callback]() {
+          callback(DataSource::read(source, allow_empty, api));
+          init_target_.ready();
```
Comment (silentdai) [inline source/common/config/datasource.h]: Warning: Take care. There is ongoing efforts on making Target remember `ready()` invoked before InitManager invokes `initialize()`. By then this `ready()` may not act as you expected

File: source/common/config/datasource.h [hunk 3]
```diff
@@ -25,6 +32,108 @@ std::string read(const envoy::api::v2::core::DataSource& source, bool allow_empt
...
+   * Create async data provider for local data source.
+   * @param manager init manager for coordinating initialization
+   * @param source data source configuration
+   * @param allow_empty if the data is allowed empty
+   * @param api "Public" API that different components use to interact with the various system
+   * abstractions
+   * @param callback data provider callback
+   * @return data provider for local data source
+   */
+  static LocalAsyncDataProviderPtr create(Init::Manager& manager,
```
Comment (silentdai) [inline source/common/config/datasource.h]: Not sure what is the value of this static method

File: source/common/config/remote_data_fetcher.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+private:
+  bool complete_{};
+
+  Upstream::ClusterManager& cm_;
+  const envoy::api::v2::core::HttpUri& uri_;
+  const std::string content_hash_;
+  RemoteDataFetcherCallback& callback_;
+
+  Http::AsyncClient::Request* request_{};
```
Comment (silentdai) [inline source/common/config/remote_data_fetcher.h]: I am not very familiar with the request_ lifetime. LGTM from the lua usage.

File: source/common/config/remote_data_fetcher.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+   */
+  void fetch();
+
+  /**
+   * Cancel the fetch.
+   */
+  void cancel();
+
+private:
+  bool complete_{};
```
Comment (silentdai) [inline source/common/config/remote_data_fetcher.h]: It seems to me `complete_==true` is always aligning with a valid request_. Is complete_ useful?

File: source/common/config/datasource.h [hunk 6]
```diff
@@ -25,6 +32,108 @@ std::string read(const envoy::api::v2::core::DataSource& source, bool allow_empt
...
+    init_target_.ready();
+  }
+
+  // Config::DataFetcher::RemoteDataFetcherCallback
+  void onFailure(Config::DataFetcher::Failure failure) {
+    if (allow_empty_) {
+      callback_(EMPTY_STRING);
+      init_target_.ready();
+    } else {
+      throw EnvoyException(
```
Comment (silentdai) [inline source/common/config/datasource.h]: You may want to rethink about whether the exception is good or not. RemoteDataFetch doesn't catch it and would omit one statement

File: source/common/config/remote_data_fetcher.cc [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+  request_ = nullptr;
+}
+
+void RemoteDataFetcher::onFailure(Http::AsyncClient::FailureReason reason) {
+  ENVOY_LOG(debug, "fetch remote data [uri = {}]: network error {}", uri_.uri(), enumToInt(reason));
+  complete_ = true;
+  callback_.onFailure(Failure::Network);
+
+  request_ = nullptr;
```
Comment (silentdai) [inline source/common/config/remote_data_fetcher.cc]: won't be executed on exception

**Comment by crazyxy** [comment]
> It is very confusing which is responsible for catching the exception RemoteAsyncDataProvider::onFailure
> Could you elaborate?

The exception will terminate either the main thread or worker thread. I think the behavior is consistent with [read](https://github.com/envoyproxy/envoy/blob/f683743ca6727d003a30c998ca617e31d9523e3d/source/common/config/datasource.cc#L9).

File: source/common/config/remote_data_fetcher.cc (hunk 7 above)
Comment (crazyxy) [reply source/common/config/remote_data_fetcher.cc]: Good catch. I will move it before callback.

File: source/common/config/remote_data_fetcher.h (hunk 5 above)
Comment (crazyxy) [reply source/common/config/remote_data_fetcher.h]: I will remove the `complete_`.

File: source/common/config/datasource.h (hunk 3 above)
Comment (crazyxy) [reply source/common/config/datasource.h]: I will remove them.

**Comment by crazyxy** [comment]
@lizan Could you please take a look? Thanks!

File: source/common/config/datasource.h [hunk 8]
```diff
@@ -25,6 +32,72 @@ std::string read(const envoy::api::v2::core::DataSource& source, bool allow_empt
...
+                                                                          source.sha256(), *this)),
+        init_target_("RemoteAsyncDataProvider", [this]() { start(); }) {
+    manager.add(init_target_);
+  }
+
+  ~RemoteAsyncDataProvider() { init_target_.ready(); }
+
+  // Config::DataFetcher::RemoteDataFetcherCallback
+  void onSuccess(absl::string_view data) {
+    callback_(std::string(data));
```
Comment (lizan) [inline source/common/config/datasource.h]: If you're constructing string from string_view anyway then probably just take a `const string&`

Comment (crazyxy) [reply source/common/config/datasource.h]: I will change the callback parameter to `const string&`.

File: source/common/config/remote_data_fetcher.h [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+ */
+class RemoteDataFetcherCallback {
+public:
+  virtual ~RemoteDataFetcherCallback() = default;
+
+  /**
+   * This function will be called when data is fetched successfully from remote.
+   * @param data remote data
+   */
+  virtual void onSuccess(const absl::string_view data) PURE;
```
[the remark this answers is not in the document]
Comment (crazyxy) [reply source/common/config/remote_data_fetcher.h]: Good catch!

File: source/common/config/datasource.h (hunk 6 above)
Comment (crazyxy) [reply source/common/config/datasource.h]: I moved the statement before the callback. I think throwing exception is consistent with current data source implementation.

**Comment by crazyxy** [comment]
> First quick skim, generally looks good and in right direction.

Thanks @lizan for reviewing! Made changes as requested.

**Comment by alyssawilk** [comment]
At some point can you do a master merge, and move your release note to 1.12.0?  Thanks!

**Comment by crazyxy** [comment]
@lizan gentle ping

File: source/common/config/datasource.h (hunk 8 above)
Comment (lizan) [reply source/common/config/datasource.h]: I meant you can make onSuccess take const string& as well.

File: source/common/config/remote_data_fetcher.cc [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+      const auto len = response->body()->length();
+      const auto content_hash =
+          Hex::encode(Envoy::Common::Crypto::Utility::getSha256Digest(*response->body()));
+
+      if (content_hash_ != content_hash) {
+        ENVOY_LOG(debug, "fetch remote data [uri = {}]: data is invalid", uri_.uri());
+        callback_.onFailure(FailureReason::InvalidData);
+      } else {
+        callback_.onSuccess(
+            absl::string_view(static_cast<char*>(response->body()->linearize(len)), len));
```
Comment (lizan) [inline source/common/config/remote_data_fetcher.cc]: response->body()->toString()

(and make onSuccess take const string&

**Comment by crazyxy** [comment]
> The code LGTM for now. The next step is add some real usage and/or integration test with xDS?

Yes. I think maybe we can apply this to WASM first as describe in this issue #7311. Meanwhile, if other components have async data source requirement, I'd like to write the first example using async data source.

**Comment by crazyxy** [comment]
@lizan, could you help merge the pr? Let me know if you have any concerns. Thanks!

**Comment by lizan** [comment]
@crazyxy Can you merge master and resolve conflicts of version_history?

**Comment by crazyxy** [comment]
> @crazyxy Can you merge master and resolve conflicts of version_history?

It's weird. Not sure why version_history conflicts. Merged upstream/master.

**Comment by lizan** [comment]
> It's weird. Not sure why version_history conflicts. Merged upstream/master.

It usually happens after new release that version_history only has a few lines for the upcoming release. so the 3-way merge fail.

**Pull request merged**